src/test/core/JobCounter_test.cpp

```diff
@@ -0,0 +1,124 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/core/JobCounter.h>
+#include <ripple/beast/unit_test.h>
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+namespace ripple {
+
+//------------------------------------------------------------------------------
+
+class JobCounter_test : public beast::unit_test::suite
+{
+    void testWrap()
+    {
+        // Verify reference counting.
+        JobCounter jobCounter;
+        BEAST_EXPECT (jobCounter.count() == 0);
+        {
+            auto wrapped1 = jobCounter.wrap ([] (Job&) {});
+            BEAST_EXPECT (jobCounter.count() == 1);
+
+            // wrapped1 should be callable with a Job.
+            {
+                Job j;
+                (*wrapped1)(j);
+            }
+            {
+                // Copy should increase reference count.
+                auto wrapped2 (wrapped1);
+                BEAST_EXPECT (jobCounter.count() == 2);
+                {
+                    // Moving from wrapped2 doesn't decrease count.
+                    auto wrapped3 (std::move(wrapped2));
+                    BEAST_EXPECT (jobCounter.count() == 3);
+                    {
+                        // An additional Job also increases count.
+                        auto wrapped4 = jobCounter.wrap ([] (Job&) {});
+                        BEAST_EXPECT (jobCounter.count() == 4);
+                    }
+                    BEAST_EXPECT (jobCounter.count() == 3);
+                }
+                BEAST_EXPECT (jobCounter.count() == 2);
+            }
+            BEAST_EXPECT (jobCounter.count() == 1);
+        }
+        BEAST_EXPECT (jobCounter.count() == 0);
+
+        // Join with 0 count should not stall.
+        jobCounter.join();
+
+        // Wrapping a Job after join() should return boost::none.
+        BEAST_EXPECT (jobCounter.wrap ([] (Job&) {}) == boost::none);
+    }
+
+    void testWaitOnJoin()
+    {
+        // Verify reference counting.
+        JobCounter jobCounter;
+        BEAST_EXPECT (jobCounter.count() == 0);
+
+        auto job = (jobCounter.wrap ([] (Job&) {}));
+        BEAST_EXPECT (jobCounter.count() == 1);
+
+        // Calling join() now should stall, so do it on a different thread.
+        std::atomic<bool> threadActive {false};
+        std::thread localThread ([&jobCounter, &threadActive] ()
+        {
+            threadActive.store (true);
+
+            // Should stall after calling join.
+            jobCounter.join();
+
+            threadActive.store (false);
+        });
+
+        // Wait for the thread to start.
+        while (threadActive.load() == false);
+        BEAST_EXPECT (jobCounter.count() == 1);
+
+        // Wait a millisecond.  The thread should still be active.
+        using namespace std::chrono_literals;
+        std::this_thread::sleep_for (1ms);
```

Comment:
It looks like we may be able to tweak this.
- The ref count on line 98 is superfluous as line 82 already checked it and it could not have possibly changed.
- The `threadActive` check on line 103 is unnecessary as the while loop would have never completed.
- The sleep does not guarantee that `join` will block before the job is destroyed. Whether the job is destroyed before or after `join` evaluates the count will have no bearing on the test result. I think your intention here was to block.

Unfortunately, we are still left with no reliable method of knowing `join` is blocking. One idea that comes to mind is adding a `JobCounter::joinable` function similar to std::thread::joinable. Taking advantage of `join` atomically releasing the lock on `wait`, `joinable` would return false after `join` blocks. It would look like this...
`        std::atomic<bool> threadExited {false};
        std::thread localThread ([&jobCounter, &threadExited] ()
        {
            // Should stall after calling join.
            jobCounter.join();
            threadExited = true;
        });

        while (jobCounter.joinable() == true);

        // Destroy the Job and expect the thread to exit (asynchronously).
        job = boost::none;
        BEAST_EXPECT (jobCounter.count() == 0);

        // Wait for the thread to exit.
        while (threadExited == false);
        localThread.join();`